pylint/checkers/variables.py

```diff
@@ -565,6 +566,8 @@ def mark_as_consumed(self, name, consumed_nodes):
         If all of the nodes for the name were consumed, delete the name from
         the to_consume dictionary
         """
+        if name not in self.to_consume:
```

Comment:
Is this change necessary because of the change to `get_next_to_consume`? I do not know why but the docstring of this method does not seem to indicate that the name not being in `self.to_consume` is an option.

I find the code of the consumption checker to be quite fragile so I'm worried this might break something that was working based on an unwritten assumption.
I don't think this change necessarily breaks anything but I wonder if this function should be called in these cases anyway or if we should have returned in an earlier method that calls `mark_as_consumed`.

Reply:
> Is this change necessary because of the change to get_next_to_consume? 

Yes, but it will not be necessary after the refactor in #5241 plus the introduction of `consumed_uncertain` in #5402. If you want to hold both PRs for the refactor, that's fine, but if you want to merge this one I will remove the unneeded block after rebasing #5402 on #5241.

The solution in these PRs is that some variables should be partially-consumed: i.e. consumed for the purposes of not being "unused" in the except/finally blocks, or after the try; but still able to emit `unused-before-assignment` if they are used in those blocks (improperly). The old pattern unconditionally marks everything as consumed before exiting the function, so without adding a new attribute like `consumed_uncertain` to store things off, this PR had to let them be marked as consumed twice and added this short-circuit.

Reply:
Let's do #5241 > #5402 > #5384 then. That seems to make the most sense 😄

I'll have another good look at #5402 after #5241 has been merged. I really want to make the consumption checker as intuitive as possible seeing the large amount of issues and false positives/negatives we currently have. I believe the code being much harder to read compared to other parts of the code are a large reason why these issues stay open. These PRs are really valuable and are a good way to start improving on it, thanks for that!